// fitnode/doctree.js
var fitnode = require('../fitnode');

/***********  根据表间关系构造树(json) **********/

var doctree = function (db, config) {
    this.db = db;
    this.config = config;
}

exports.doctree = doctree;

doctree.prototype = {
    constructor: doctree,
    middleware: {
        az: {
            dir: function (type, subtype, parent_attr) {
                var data = [];
                var letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
                for (c in letters) {
                    var l = { filter: "[" + letters[c].toUpperCase() + letters[c].toLowerCase() + "]", text: letters[c], parent_type: type,
                        sub_type: subtype, middleware: 'az'
                    };
                    for (var i in parent_attr)
                        if (i && !l[i]) l[i] = parent_attr[i];
                    data.push(l);
                }
                return data;
            },
            children: function (parent_attr) {
                return [{ type: parent_attr['sub_type'], query: { TableName: /$t[Aa]/}}];
            }
        }
    },

    get: function (attr, callback) {
        var type = attr['type'];
        var c = this.config[type];
        var data = [];
        if (attr['middleware']) {
            c.children = this.middleware[attr.middleware].children(attr);
        }
        new new fitnode.seq_asyncArray(
            function (item, data, callback) {
                if (item.middleware) {
                    data = this.middleware[item.middleware].dir(type, item.type, attr);
                    callback(data);
                }
                else {
                    if (item.type) {
                        var cc = this.config[item.type];
                        console.log(item.query);
                        this.db.collection(cc.table).find(item.query).toArray(function (err, rows) {
                            for (var i in rows)
                                data.push({ text: rows[i][cc.text] });
                        });
                    }
                }
            },
            c.children,
            function () { callback(data); }
        ).exec();
    }
}
