// fitnode/index.js
var snsauth = function (db, config) {
    this.db = db;
    this.config = config;
    //console.log(config.roles);
}

exports.snsauth = snsauth;

snsauth.prototype = {
    constructor: snsauth,
    limit: function (query, role_code, account, rel_account) {
        var auth = this;
        this.db.collection('user_relation').find({
            //account: account, rel_account: rel_account, //以后加
            "roles.code": role_code
        }).toArray(function (err, items) {
            var w = [];
            if (items.length == 1)
                for (var i in items[0].roles) {
                    var role = items[0].roles[i];
                    
                    var funcs = auth.config.sysroles[role.code].funcs;
                    for (var j in funcs) {
                        var func = auth.config.sysfuncs[funcs[j]];
                        if (func.doc)
                            w.push(auth.db.ObjectID(role.doc_id));
                    }
                }
            query["_id"] = {"$in": w};
            console.log(query);
            return query;
        });
    },
    demo_data: function () {
        this.db.collection('user_define_roles').remove().insert({
            account: '4f61a6b86e435c09b885ad0a', name: '基友', sysroles: ['friend', 'dev']
        });

        this.db.collection('user_relation').remove().insert({
            account: '4f61a6b86e435c09b885ad0a',
            rel_account: '4f619a9365631f0d14dc3b2c', //lobby@163.com
            roles: [
                { code: 'friend', doc_id: null },
                { code: 'dev', doc_id: '4f619a9365631f0d14dc3b2d' },
                //{ id: 'xxx', doc_id: null }
            ]
        });
    }
};
